spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.handler;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.aggregator.CorrelationStrategy;
+import org.springframework.integration.aggregator.HeaderAttributeCorrelationStrategy;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+
+/**
+ * A pair of message handlers; the first suspends the thread until the second receives
+ * a corresponding message.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class SuspendingMessageHandler extends AbstractReplyProducingMessageHandler {
```

Comment:
Well, even if `Suspending` word is more closer to the implementation of the class, but in face of Java Config, I'd prefer now to go ahead with symmetrical names for XML component, as well as for their classes on the background.